src/sage/rings/polynomial/multi_polynomial_ideal.py

```diff
@@ -4973,7 +4987,10 @@ def reduce(self, f):
             pass
 
         gb = self.groebner_basis()
-        return f.reduce(gb)
+        R = self.ring()
+        if not R.has_coerce_map_from(parent(f)):
+            raise TypeError(f"element belong to {parent(f)}, cannot coerce to {R}")
+        return R(f).reduce(gb)
```

Comment:
```suggestion
        f = self.ring().coerce(f)
        return f.reduce(gb)
```
I think this will be sufficient.